### PR TITLE
Playwright tests - Parent DAO Creation

### PR DESCRIPTION
**Pull request by SudoPumpkin**
## Description

- Several selectors have been updated in the Fractal app to aid in solid test results. 
- A new test to create a 1:1 Gnosis Parent DAO has been created. 
- Test 002 has been updated with a delay. 

## Notes

Playwright needs to be listed in the dependencies of the package.json due to how Playwright references imports. That's why it's been added. 

## Issue (if applicable)

A manual delay was added to allow the app and local wallet to be loaded with multiple workers running at the same time.

## Testing

`npm run tests
`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -23,6 +23,7 @@
     "ethers": "^5.7.0",
     "i18next": "^21.9.2",
     "i18next-browser-languagedetector": "^6.1.5",
+    "@playwright/test": "^1.26.0",
```
Comment (adamgall) [inline package.json:26]: Please put this in alphabetical order in the dependency list

**Review by adamgall — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (SudoPumpkin) [reply package.json]: Done. =)

File: src/components/ui/Header/MenuItems.tsx [hunk 2]
```diff
@@ -108,14 +116,22 @@ function ItemContent({
...
+  testId,
+}: ActionMenuItem) {
   const { t } = useTranslation('menu');
   if (!isVisible) {
     return null;
   }
   return (
     <Menu.Item>
       <button
+        data-testid={testId}
```
Comment (adamgall) [inline src/components/ui/Header/MenuItems.tsx:134]: Is there a reason why this attribute isn't just called `id`?

File: src/components/ui/Header/MenuItems.tsx [hunk 3]
```diff
@@ -147,11 +164,14 @@ function LinkItem({ titleKey: title, link, Icon }: LinkMenuItem) {
...
 
-function LinkItemInternal({ titleKey: title, link, Icon }: LinkMenuItem) {
+function LinkItemInternal({ titleKey: title, link, Icon, testId }: LinkMenuItem) {
   const { t } = useTranslation('menu');
   return (
     <Menu.Item>
-      <Link to={link}>
+      <Link
+        to={link}
+        data-testid={testId}
```
Comment (adamgall) [inline src/components/ui/Header/MenuItems.tsx:173]: Is there a reason why this attribute isn't just called `id`?

**Review by adamgall — COMMENTED** [review]
Left some comments

File: src/components/ui/Header/MenuItems.tsx (hunk 2 above)
Comment (adamgall) [reply src/components/ui/Header/MenuItems.tsx:134]: i got an answer -- because sometimes React overwrites the `id` attribute

File: src/components/ui/Header/MenuItems.tsx [hunk 4]
```diff
@@ -129,11 +145,12 @@ function ActionItem({ titleKey: title, className, action, Icon, isVisible }: Act
   );
 }
 
-function LinkItem({ titleKey: title, link, Icon }: LinkMenuItem) {
+function LinkItem({ titleKey: title, link, Icon, testId }: LinkMenuItem) {
   const { t } = useTranslation('menu');
   return (
     <Menu.Item>
       <a
+        data-testid={testId}
```
[the remark this answers is not in the document]
Comment (adamgall) [reply src/components/ui/Header/MenuItems.tsx:153]: i got an answer -- because sometimes React overwrites the `id` attribute

File: src/components/ui/Header/MenuItems.tsx (hunk 3 above)
Comment (adamgall) [reply src/components/ui/Header/MenuItems.tsx:173]: i got an answer -- because sometimes React overwrites the `id` attribute

File: playwright.config.ts [hunk 5]
```diff
@@ -13,7 +13,7 @@ const config: PlaywrightTestConfig = {
      * Maximum time expect() should wait for the condition to be met.
      * For example in `await expect(locator).toHaveText();`
      */
-    timeout: 10000
+    timeout: 15000
```
[the remark this answers is not in the document]
Comment (adamgall) [reply playwright.config.ts:16]: I got an answer, it's because stuff sucks and sometimes you just gotta wait a lil longer

File: tests/002confirmWalletConnect.spec.ts [hunk 6]
```diff
@@ -1,24 +1,31 @@
 import { expect, test } from '@playwright/test';
```
Comment (herbig) [inline tests/002confirmWalletConnect.spec.ts:1]: not a change from this PR, but it might be good to rename this `002confirmWalletConnected`, just so it's not confused with being a test for WalletConnect the wallet brand.

File: page-objects/components/Navbuttons.ts [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    this.localWallet = page.locator(
+      '#WEB3_CONNECT_MODAL_ID div.web3modal-provider-name:has-text("Local Node")'
+    );
+    this.next = page.locator('button:has-text("* * *Next")');
+    this.deploy = page.locator('button:has-text("* * *Deploy")');
+  }
+
+  async clickOnButton(buttonName: any) {
+    switch (buttonName) {
+      case 'Return to Homepage':
```
Comment (herbig) [inline page-objects/components/Navbuttons.ts:59]: Is there a reason for having these designated with string descriptions rather than having a separate method for each clickable item on screen?

I feel like having `clickReturnToHomePage();`, `clickHeaderConnectWallet();`, etc, would be cleaner than a large switch statement with new strings you need to remember when writing the tests.   You also get autocomplete this way.

File: tests/003createParentDao.spec.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  let notifications: Notifications;
+
+  test.beforeEach(async ({ page }) => {
+    daoSafe = new DaoSafe(page);
+    homePage = new HomePage(page);
+    navButtons = new Navbuttons(page);
+    notifications = new Notifications(page);
+    inputFields = new InputFields(page);
+    homePage.visit();
+    await delay(3000);
```
Comment (herbig) [inline tests/003createParentDao.spec.ts:23]: can we put this delay in the `visit();` methods of each page instead?

**Review by Da-Colon — APPROVED** [review]
Approved pending Mike's comments

File: tests/002confirmWalletConnect.spec.ts (hunk 6 above)
Comment (SudoPumpkin) [reply tests/002confirmWalletConnect.spec.ts:1]: I can rename this to ```002connectToWallet```, does this work?

File: tests/003createParentDao.spec.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    notifications.closeButton('Close Audit Message');
+    await page.waitForLoadState();
+    await delay(1500);
```
[the remark this answers is not in the document]
Comment (SudoPumpkin) [reply tests/003createParentDao.spec.ts:26]: If the current page has already reached the required state, it resolves the wait immediately. So here no particular state is defined so it's just waiting until load is completed to continue. I can remove this and increase the delay time to '2000'. I had this here to negate a minor amount of ms in wait time compared to a hard increased delay wait. I can totally just add another 500ms(0.5secs) to the present '1500'(1.5secs) to make it '2000'(2secs) and yoink line 25.

File: page-objects/components/Navbuttons.ts (hunk 7 above)
Comment (SudoPumpkin) [reply page-objects/components/Navbuttons.ts:59]: With similar action components this is a common method for e2e test scripting. It's easier for code readability.  I can change it however to be similar to HomePage.ts.

File: tests/003createParentDao.spec.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (SudoPumpkin) [reply tests/003createParentDao.spec.ts:26]: Actually I am leaving line 25 in there. I removed line 25 locally, and when the tests ran in parallel multiple times they become flaky. A test will random fail in different browser.

File: tests/002confirmWalletConnect.spec.ts (hunk 6 above)
Comment (herbig) [reply tests/002confirmWalletConnect.spec.ts:1]: yeah that sounds good, I like the distinction

File: page-objects/components/Navbuttons.ts (hunk 7 above)
Comment (herbig) [reply page-objects/components/Navbuttons.ts:59]: cool cool, sounds good, I defer to you on that, just curious what's common practice

File: tests/003createParentDao.spec.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (herbig) [reply tests/003createParentDao.spec.ts:26]: sounds good.  I'd just like to better understand why the hardcoded waits everywhere are necessary, and what the difference is when running locally

File: tests/003createParentDao.spec.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  });
+
+  test('Create 1:1 Gnosis Parent DAO', async ({ page }) => {
+    navButtons.clickOnButton('Create a Fractal - Button');
+
+    /* Check URL to make sure navigation is correct. */
+    await expect(page).toHaveURL('http://localhost:3000/#/daos/new');
+
+    /* Input a Fractal name */
+    inputFields.fillField('Insert Fractal Name');
```
[the remark this answers is not in the document]
Comment (herbig) [reply tests/003createParentDao.spec.ts:45]: that makes sense!

File: tests/003createParentDao.spec.ts (hunk 8 above)
Comment (SudoPumpkin) [reply tests/003createParentDao.spec.ts:23]: No idea, but it can't find the menu 'connect' option without it.

File: tests/003createParentDao.spec.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (SudoPumpkin) [reply tests/003createParentDao.spec.ts:26]: Locally it runs with up to 8 workers in parallel while in CI it runs 1 worker. These 8 workers are all running at the same time. Each browser behaves a little bit differently from one another with load times of content with the Fractal app. Now, let's add the local wallet connecting to the app and local blockchain on top of that varied browser behavior. 
The local provider checks to see if the network is active before it connects. When the app loads, it loads this twice(because react in dev). In the middle of this async call the test goes and clicks the connect button. The promise is then returned and interrupts the other state setting. The other 'fallback' provider is not async, there is are promises.

Long story short, sometimes the wallet isn't ready for Playwright to click 'connect' and it 1/2 connects. These hard delays force Playwright to wait for all the pieces to be loaded.
